Highlight Placement Tool

ericmehl opened the pull request:
This adds a variation to the Light Position Tool that moves the selected light such that it creates a specular highlight at the target point.

The test for the placement calculation is failing currently, but I think that may be a matter of a bad test rather than an actual failure. It's failing in the orientation calculation and if anything, this seems to be better at maintaining the orientation than the shadow tool, so I think it could be a matter of pesky equivalent Euler rotations.

It's a bit noisy when dragging but clicking individually seems to be mostly pretty good though not 100% perfect in my testing.

### Checklist ###

- [X] I have read the [contribution guidelines](https://github.com/GafferHQ/gaffer/blob/main/CONTRIBUTING.md).
- [X] I have updated the documentation, if applicable.
- [X] I have tested my change(s) in the test suite, and added new test cases where necessary.
- [X] My code follows the Gaffer project's prevailing coding style and conventions.

ericmehl commented:
No Euler angle shenanigans, just a regular old vector reversal when computing the desired orientation value. Fixed and squashed into https://github.com/GafferHQ/gaffer/pull/5630/commits/05716fff74260a1f408445b4331a43a97ea2a8b6.

johnhaddon reviewed (commented):
Thanks Eric! When this works, it works great! I am getting some less-than-accurate placements at times though, some of which even seem so far off that I'm not sure precision in the normal is the only problem (maybe I'm hitting the visualisation somehow?). We'll need a bit more iteration here I think - maybe something @danieldresser-ie might be able to assist with?

One other user-facing improvement we're in need of is to update the tooltip overlay when switching modes. I think the easiest way to do that would be to have `_TargetTipWidget` listen to `metadataChangedSignal()` and then emit that signal from `plugSet()`.

This isn't strictly a topic for this PR, but I am still finding myself occasionally accidentally repositioning geometry rather than lights with this thing. Since the placement modes are so intrinsic to lights, I think it might be worth just disabling the tool for any other types of selection. But I wonder if this is just me?

ericmehl commented:
I haven't addressed the comments yet, but the latest push includes a much more accurate normal computation, and therefore a smoother reflection placement as you drag.

There's one nagging precision issue I'll point out inline that can make the highlight placement even better, but it's a bit tricky to sort out where to make a fix.

ericmehl commented on src/GafferSceneUI/SceneGadget.cpp:
```diff
@@ -512,13 +502,15 @@ bool SceneGadget::objectAt( const IECore::LineSegment3f &lineInGadgetSpace, Gaff
 
+	auto viewportGadget = ancestor<ViewportGadget>();
+	const V2f rasterPosition = viewportGadget->gadgetToRasterSpace( lineInGadgetSpace.p1, this );
+	const LineSegment3f offsetLineInGadgetSpace = viewportGadget->rasterToGadgetSpace( rasterPosition + V2f( 0.5f ), this );
```
Doing this gadget -> raster + offset -> gadget space conversion adds a noticeable amount of accuracy loss. If I save out the normals for a viewport using this method, and by passing in an already offset `lineInGadgetSpace` (so omitting the extra conversion), the pre-offset one is much better.

I haven't looked at improving precision yet, it could be there's some wins to be had by doing computation in double precision but still passing single precision floats back and forth.

danieldresser-ie replied:
I don't fully understand the context here, but why wouldn't we want to pass in an already offset lineInGadgetSpace? I'm assuming this is so that we get a vector that points through the center of the pixel, which seems most correct? Is the reason not to change this just because it would be a compatibility break?

ericmehl replied:
Yes, we need a vector that passes through the center of the pixel. We're only given a `GafferUI::ButtonEvent` object in the click handlers, like https://github.com/GafferHQ/gaffer/blob/957ab7fd741232bdcea701ec335d78bc469164c3/src/GafferSceneUI/LightPositionTool.cpp#L770

( `DragDropEvent` derives from `ButtonEvent` ) `ButtonEvent` stores the ray but not the pixel coordinate. So as things are we'll need to do some kind of conversion I guess.

danieldresser-ie replied:
I guess it would take a fair bit of an audit to identify everywhere where the ray stored in ButtonEvent is used ... but it seems unlikely that it would ever be desirable to have that ray point to a corner instead of the center?

johnhaddon replied:
I think Daniel's right here - the ray should already be pointing through the centre of the pixel, long before it ends up in `objectAt()`. Exactly where we put that adjustment is the question. This is the current pipeline as I understand it :

1. We get an event from Qt. From that we get an integer pixel position relative to the widget's origin. Qt does provide a floating point accessor for the position (and in Qt 6 it seems _only_ that), but that is returning integer values too anyway.
2. We convert this into our `ButtonEvent.line`. All our events are floating point, and specify a ray (`.line`) instead of a 2d position, because we want to use them both in Widgets (2d) and Gadgets (3d). At this point our ray points straight into the screen, orthographic-like. But we haven't added a `0.5` offset here.
3. GadgetWidget passes these events to its internal ViewportGadget.
4. ViewportGadget transforms the origin of the line from raster space into 3d space using the camera projection via `rasterToGadgetSpace()`. From there it transforms into the space of whichever gadget it dispatches the event to. That's all done in `eventToGadgetSpace()`.

The places we could put the offset are 2 and 4. I don't think it belongs in `ViewportGadget::rasterToGadgetSpace()` itself because that takes floating point coordinates (if it took integers, that would be a different matter). I think most logically it belongs at 2, the first place we convert from integer coordinates to floating point. That affects a much greater amount of code (all Widgets potentially), so is the riskier choice - I suspect at least one will be thinking in integers. The alternative is 4, applying the offset before calling `rasterToGadgetSpace()`. I don't think it's the _right_ place, but it has the benefit of limiting the knock on effects to only Gadgets.

Since 1.4 is so close, I'm feeling fairly inclined to go with the slightly riskier but more correct option, and just not put it in 1.3.x. But I'm open to other opinions...

ericmehl commented on src/GafferSceneUI/LightPositionTool.cpp:
```diff
@@ -679,15 +710,32 @@ void LightPositionTool::updateHandles( float rasterScale )
 	const V3f p = transform.translation();
-	const Line3f handleLine( shadowTarget.value(), shadowPivot.value() );
+
+	Line3f handleLine;
```
Fixed in https://github.com/GafferHQ/gaffer/pull/5630/commits/a17496c4a24557ca7d1d5a75bd6ecdf86fd33dc2

ericmehl commented on include/GafferSceneUI/LightPositionTool.h:
```diff
@@ -62,10 +62,22 @@ class GAFFERSCENEUI_API LightPositionTool : public GafferSceneUI::TransformTool
 		GAFFER_NODE_DECLARE_TYPE( GafferSceneUI::LightPositionTool, LightPositionToolTypeId, TransformTool );
 
+		Gaffer::IntPlug *toolTypePlug();
+		const Gaffer::IntPlug *toolTypePlug() const;
```
Fixed in https://github.com/GafferHQ/gaffer/pull/5630/commits/7b8f0c043265bbd6bb9463407f64913419076712

ericmehl commented on python/GafferSceneUI/LightPositionToolUI.py:
```diff
@@ -57,4 +57,30 @@
+			"""
+			The method to use for placing the light.
+
+			- Shadow : Place the light so that it casts a shadow from the pivot point
```
Fixed in https://github.com/GafferHQ/gaffer/pull/5630/commits/e353cd00b54518c153c9ac67d55563ba19bc9fe5

ericmehl commented on python/GafferUI/Viewer.py:
```diff
@@ -234,17 +234,16 @@ def __keyPress( self, widget, event ) :
-					t["orientation"],
-					presets[ ( presets.index( orientation ) + 1 ) % len( presets ) ]
-				)
+			for p in t :
```
Fixed in https://github.com/GafferHQ/gaffer/pull/5630/commits/0d3cd30f72eff4483f6ca9cd538e5fb3e9e8083e

ericmehl commented on python/GafferUI/Viewer.py:
```diff
@@ -234,17 +234,16 @@ def __keyPress( self, widget, event ) :
-				)
+			for p in t :
+				cycleKey = Gaffer.Metadata.value( p, "viewer:cyclePresetShortcut" )
+				if t["active"].getValue() and event.key == cycleKey :
```
Fixed in https://github.com/GafferHQ/gaffer/pull/5630/commits/6c839788e5a69c04beff8369a23635f6f859bade

ericmehl commented on include/GafferSceneUI/SceneGadget.h:
```diff
@@ -182,6 +182,11 @@ class GAFFERSCENEUI_API SceneGadget : public GafferUI::Gadget
+		/// Returns the approximate gadget space normal of the frontmost object intersecting
+		/// the specified line through gadget space. Returns true on success and false
+		/// if there is no such object.
+		bool normalAt( const IECore::LineSegment3f &lineInGadgetSpace, Imath::V3f &gadgetNormal ) const;
```
We talked offline about using `std::optional<V3f>` which is done in https://github.com/GafferHQ/gaffer/pull/5630/commits/0854f737b903ecdeda85d02e3a49aa5ddbb31252.

ericmehl commented on python/GafferSceneUITest/SceneGadgetTest.py:
```diff
@@ -106,6 +107,19 @@ def assertObjectAt( self, gadget, ndcPosition, path ) :
+		self.assertEqual( r, result )
+		for i in range( 0, 3 ) :
+			with self.subTest( "{} Axis".format( ["X", "Y", "Z"][i] ), i = i ) :
+				self.assertAlmostEqual( n[i], normal[i], places = 1 )
```
That does make for a much more intuitive understanding of how accurate the calc / tests are. I got it down to 5 degrees max error (around 1-2 for most samples) which isn't quite as low as I'd like, but it seems pretty accurate in practice.

The overall accuracy boost, including fixing this note, is in https://github.com/GafferHQ/gaffer/pull/5630/commits/0854f737b903ecdeda85d02e3a49aa5ddbb31252

ericmehl commented on python/GafferSceneUITest/SceneGadgetTest.py:
```diff
@@ -675,6 +689,58 @@ def testSetRenderer( self ) :
+		self.assertNormalAt( sg, imath.V2f( 0.5 ), imath.V3f( 0, 0, 1.0 ), True )
+		origin = imath.V3f( 0, 0, 2 )
+		direction = imath.V3f( 0.5, 0, -2 ).normalized()
+		t = self.__rayCircleIntersectionT( origin, direction, imath.V3f( 0.0 ), 1.0 )
```
Fixed in https://github.com/GafferHQ/gaffer/pull/5630/commits/0854f737b903ecdeda85d02e3a49aa5ddbb31252

johnhaddon replied:
Sorry, my original comment wasn't clear - I was questioning whether `__rayCircleIntersection()` should have been called `__sphereIntersection()`.

johnhaddon replied:
Looks like this comment is still outstanding?

ericmehl commented on python/GafferSceneUITest/SceneGadgetTest.py:
```diff
@@ -675,6 +689,58 @@ def testSetRenderer( self ) :
+		direction = imath.V3f( 0.5, 0, -2 ).normalized()
+		t = self.__rayCircleIntersectionT( origin, direction, imath.V3f( 0.0 ), 1.0 )
+		sphereIntersection = origin + direction * t
+		self.assertNormalAt( sg, imath.V2f( 0.75, 0.5 ), sphereIntersection, True )
```
Fixed in https://github.com/GafferHQ/gaffer/pull/5630/commits/0854f737b903ecdeda85d02e3a49aa5ddbb31252

ericmehl commented on src/GafferSceneUI/SceneGadget.cpp:
```diff
@@ -656,6 +656,86 @@ size_t SceneGadget::objectsAt(
+		// where each pair has the same object ID as the center pixel.
+
+		auto viewportGadget = ancestor<ViewportGadget>();
+		const V2f rasterPosition = viewportGadget->gadgetToRasterSpace( lineInGadgetSpace.p1, this );
```
After much poking and prodding of the numbers, that turned out to be the solution. Part of that was using a `box` filter for Arnold's depth buffer so we're getting the average of the pixel depth samples instead of the minimum.

Fixed in https://github.com/GafferHQ/gaffer/pull/5630/commits/0854f737b903ecdeda85d02e3a49aa5ddbb31252

ericmehl commented on src/GafferSceneUI/SceneGadget.cpp:
```diff
@@ -656,6 +656,86 @@ size_t SceneGadget::objectsAt(
+			gadgetNormal = ( topP - centerP ).cross( leftP - centerP ).normalized();
+			return true;
+		}
+		else
```
Fixed in https://github.com/GafferHQ/gaffer/pull/5630/commits/75ead09ddf23d6901455f06cab634d4eae99c946

ericmehl commented on src/GafferSceneUI/SceneGadget.cpp:
```diff
@@ -656,6 +656,86 @@ size_t SceneGadget::objectsAt(
+			topP
+		);
+
+		if( leftHit && topHit && leftPath == topPath )
```
Fixed in https://github.com/GafferHQ/gaffer/commit/75ead09ddf23d6901455f06cab634d4eae99c946

ericmehl commented on include/GafferSceneUI/LightPositionTool.h:
```diff
@@ -129,11 +139,25 @@ class GAFFERSCENEUI_API LightPositionTool : public GafferSceneUI::TransformTool
+			// The `ToolType::Shadow` tracks this naturally with the `TargetMode::Pivot`,
+			// but `ToolType::Highlight` needs this extra state since it does not
+			// use `TargetMode::Pivot`.
+			PhantomPivot,
```
Fixed in https://github.com/GafferHQ/gaffer/pull/5630/commits/3b0718636b904ef755266efefa00d86616953d4c

ericmehl commented on src/GafferSceneUI/LightPositionTool.cpp:
```diff
@@ -814,7 +843,14 @@ bool LightPositionTool::keyPress( const KeyEvent &event )
+			}
+			else
+			{
+				setTargetMode( TargetMode::PhantomPivot );
```
Same as above, fixed in https://github.com/GafferHQ/gaffer/pull/5630/commits/3b0718636b904ef755266efefa00d86616953d4c

ericmehl commented on src/GafferSceneUI/LightPositionTool.cpp:
```diff
@@ -490,20 +495,20 @@ class DistanceHandle : public Handle
+				if( !m_requiresPivot || m_shadowPivot )
+				{
+					IECoreGL::GroupPtr lineGroup = new IECoreGL::Group;
+						lineGroup->addChild(
```
Fixed in https://github.com/GafferHQ/gaffer/pull/5630/commits/edd75169cb0eb4e073096a728433e994f2588a54

ericmehl commented on src/GafferSceneUI/LightPositionTool.cpp:
```diff
@@ -911,6 +918,11 @@ void LightPositionTool::plugSet( Plug *plug )
+	{
+		auto h = static_cast<DistanceHandle *>( m_distanceHandle.get() );
+		h->setRequiresPivot( toolTypePlug()->getValue() == (int)ToolType::Shadow );
+	}
```
Yes we do! Done in https://github.com/GafferHQ/gaffer/pull/5630/commits/4867048f62eb25819b100fb7bcbaf1d5001dc75f

ericmehl commented on src/GafferSceneUI/LightPositionTool.cpp:
```diff
@@ -490,20 +495,20 @@ class DistanceHandle : public Handle
-				);
-
-				group->addChild( lineGroup );
+				if( !m_requiresPivot || m_shadowPivot )
```
Fixed in https://github.com/GafferHQ/gaffer/pull/5630/commits/56f1586be459b4cab53be1c304f810f74f1dcbf8

ericmehl commented:
Alrighty, I have addressed the feedback and noted their corresponding commits above. I also added https://github.com/GafferHQ/gaffer/pull/5630/commits/fad39307554d0d2493ed57f07767b16a736880d3 where we skip a couple tests because the gcc9 container has a Mesa OpenGL implementation that doesn't offer floating point depth buffers that we now rely on.

I'm looking at the last test failures now, where a fragment shader is not finding `geometryP`.

johnhaddon commented on python/GafferSceneUITest/SceneGadgetTest.py:
```diff
@@ -675,6 +691,76 @@ def testSetRenderer( self ) :
+			rasterPosition = imath.V2f( int( random.random() * vpSize.x ), int( random.random() * vpSize.y ) )
+			gadgetLineCenter = vp.rasterToGadgetSpace( rasterPosition + imath.V2f( 0.5 ), sg )
+
+			# Calculate the eye ray through `rasterPosition`, flipping the Y coordinate
+			# so +Y is going up. We use this instead of `gadgetLineCenter` to avoid the
+			# precision issues `rasterToGadgetSpace()` introduces.
+			vpPosition = ( 2.0 * imath.V2f( rasterPosition.x + 0.5, 500.0 - ( rasterPosition.y + 0.5 ) ) ) / imath.V2f( vpSize ) - imath.V2f( 1.0 )
+			rayDirection = ( imath.V3f( vpPosition.x, vpPosition.y, 0.0 ) - cameraOrigin ).normalized()
```
Can you clarify what precision issues `rasterToGadgetSpace()` is introducing? I has assumed the precisions problems were usually from the reverse - `gadgetToRasterSpace()`. Minimally, I'd like to come out of this PR with a clear understanding of what future improvements we'd like to make, which I thought were primarily about handling the `0.5` offset more centrally.

johnhaddon replied:
Even if `rasterToGadgetSpace()` was imprecise, I guess I also don't understand how comparing those results with a more accurate formulation would help.

ericmehl replied:
Thanks for nudging on that one, it turns out using `gadgetLineCenter` in `__rayCircleIntersection` works after all. I was probably trying to fix a deeper problem than that precision, fixed whatever that problem was then didn't try reverting back the custom camera ray calculation. I'll have a fixup for that shortly.

ericmehl replied:
Switched to using `gadgetLineCenter` in https://github.com/GafferHQ/gaffer/pull/5630/commits/6a5eb3719f25b0ca6f0fa65d5f9df83358ef2f05

johnhaddon commented on src/GafferSceneUI/SceneGadget.cpp:
```diff
@@ -656,6 +643,71 @@ size_t SceneGadget::objectsAt(
+
+	if( objectAt( viewportGadget->rasterToGadgetSpace( centerRasterP, this ), centerPath, centerP ) )
+	{
+		// Find the normal of the first triangle formed by `pixelPosition` and
+		// left + top, right + top, left + bottom or right + bottom depth samples
+		// where each pair has the same object ID as the center pixel.
```
I think this comment is out of date now, since we no longer only take the first?

ericmehl replied:
Yes it is, fixed in https://github.com/GafferHQ/gaffer/pull/5630/commits/4e0abf9ee03fba8599f5047db82b7299a833075d

johnhaddon reviewed (commented):
Thanks Eric!

I've been through and resolved everything from my previous review, apart from a couple of small queries. Couple of queries on this one too, and a couple of small useability improvements I just noted :

- The tooltip needs to change when changing between modes - currently we're still saying "Click to place pivot" in highlight mode.
- We're still showing the pivot cursor when holding down Shift in highlight mode. Since pivots are irrelevant in that mode, we should avoid giving the impression that they're settable.

When you get in, let's chat about whether we want to do those before or after a beta release. Either way, I think this is ready for squashing...

Cheers...
John

ericmehl commented on python/GafferSceneUITest/SceneGadgetTest.py:
```diff
@@ -675,6 +691,76 @@ def testSetRenderer( self ) :
+
+		for i in range( 100 ) :
+			# `rasterPosition` from the top-left corner of the viewport, +Y going down
+			rasterPosition = imath.V2f( int( random.random() * vpSize.x ), int( random.random() * vpSize.y ) )
+			gadgetLineCenter = vp.rasterToGadgetSpace( rasterPosition + imath.V2f( 0.5 ), sg )
```
Yes, that's exactly right. Because `normalAt()` snaps the coordinate to the nearest `x.5` value, we'd be comparing a ray in `__rayCircleDirection()` that could be up to half a pixel width, in gadget space, off from what `normalAt()` will be working with.

I guess since `normalAt()` accepts a float value, it could be a little misleading to someone thinking that the ray they supply is the exact ray for which the normal is calculated?

ericmehl replied:
Changed to using `gadgetLineCenter` in https://github.com/GafferHQ/gaffer/pull/5630/commits/6a5eb3719f25b0ca6f0fa65d5f9df83358ef2f05

ericmehl commented on src/GafferSceneUI/SceneGadget.cpp:
```diff
@@ -656,6 +643,71 @@ size_t SceneGadget::objectsAt(
+
+	V2f centerRasterP = viewportGadget->gadgetToRasterSpace( lineInGadgetSpace.p0, this );
+	centerRasterP.x = (float)( (int)centerRasterP.x ) + 0.5f;
+	centerRasterP.y = (float)( (int)centerRasterP.y ) + 0.5f;
```
We don't actually want to add the 0.5 offset here, right? It's added in `objectAt()`, so this would make it a full pixel offset?

ericmehl replied:
I'm pretty confident that should be removed, so I did so in https://github.com/GafferHQ/gaffer/pull/5630/commits/5b3a86f0c6f30e29c2d422c7897d1861af5b5329. I can always drop the commit if I have a case of Friday afternoon mind mush.

johnhaddon replied:
My Thursday morning mind mush concurs with your Friday afternoon mind mush. Thanks for spotting!

ericmehl commented:
I added the to-do note to `Widget.__widgetSpaceLine()` in https://github.com/GafferHQ/gaffer/pull/5630/commits/20343526f18932671c2ef099d279f5c791546153.

I also added one comment about what value to snap to in `normalAt()`. I think after that it should be ready for a squash and merge!

johnhaddon commented:
Thanks Eric, and sorry for the delay in getting back to this. I've resolved my outstanding comments where I can, but I think a few have slipped through the cracks :

- We need to update the tooltip overlay when switching modes. I suggested one way of doing that above, but perhaps an even easier one is for `_TargetTipWidget` to connect to `Tool.plugSetSignal()` and update whenever it is triggered (and then register dynamic metadata for LightPositionTool that queries the mode).
- __rayCircleIntersection() -> __sphereIntersection().
- We're still showing the pivot cursor when holding down Shift in highlight mode. Since pivots are irrelevant in that mode, we should avoid giving the impression that they're settable.

Cheers...
John

ericmehl commented:
My last two commits change the 0.5 pixel offset calculation so that it relies on the offset being done in `Widget.py` as we did in https://github.com/GafferHQ/gaffer/pull/5712. Accuracy still looks good with OpenGL, Arnold and Cycles 👍 

@murraystevenson now might be a good time to do some hopefully final testing, I believe everything is in place now.

ericmehl commented:
To summarize the new fixups :

- https://github.com/GafferHQ/gaffer/pull/5630/commits/d5a0d8ee23e0d7a6834e14e5ecbcc552af740b1d : Changes the test helper method to `__raySphereIntersection`
- https://github.com/GafferHQ/gaffer/pull/5630/commits/b79ccb460ae4ecdb26341bbcac11d7f3600fffeb : Hides the cursor when using the highlight placement mode and the pivot modifier key is held
- https://github.com/GafferHQ/gaffer/pull/5630/commits/7689140c079fac45fc303b5e18b6f69d1d845a34 : Makes the tool tip dynamic to the shadow / highlight mode
- https://github.com/GafferHQ/gaffer/pull/5630/commits/a55e25bb6627cc73d121a28deee5ac66c5c9b2b6 : Reverts the commit adding customizable filter widths. Using a filter width of 1.0 instead of the default 2.0 seemed to make Arnold less accurate.
- https://github.com/GafferHQ/gaffer/pull/5630/commits/90e9639d7987c1558d8b7fefbaa550e788a876d9 : Fixed the use of the offset pixel location to also be used by the output buffer (though this will immediately be superseded by the next commit)
- https://github.com/GafferHQ/gaffer/pull/5630/commits/f918f5c84dc58597b7f09787852bc30646d39278 : Uses the `Widget.py` offset from #5712
- https://github.com/GafferHQ/gaffer/pull/5630/commits/bc5278baf55981b41c256f23cb9b82c89f8c0244 : Snaps to the nearest `x.5` value now that we're not offsetting after the snapping

murraystevenson commented:
Thanks Eric! Accuracy is looking good to me. The drag experience with Arnold could still do with some improvement, but Arnold click placement works great and drag placement seems to be working quite well in both OpenGL and Cycles so this seems worth getting out into a 1.4 beta for some wider testing...

johnhaddon commented on src/GafferSceneUI/SceneGadget.cpp:
```diff
@@ -657,8 +657,8 @@ std::optional<V3f> SceneGadget::normalAt( const IECore::LineSegment3f &lineInGad
 	// incorrect. We know that the incoming `lineInGadgetSpace` has been calculated from
 	// an integeer pixel value so we snap to the nearest integer as a corrective.
```
integeer -> integer. Although it might make more sense to talk about pixel centres than integers now :

> We expect that `lineInGadgetSpace` corresponds to the centre of a pixel, so we snap to the nearest center as a corrective.

johnhaddon reviewed (commented):
Thanks Eric! Accuracy looking good to me too. I've made a few comments where I think we can improve the clarity of the result after some of the fixups - hopefully they make sense.

Let's get everything squashed and ready to go into the next beta...nearly there!

ericmehl commented on src/GafferSceneUI/SceneGadget.cpp:
```diff
@@ -506,16 +506,15 @@ bool SceneGadget::objectAt( const IECore::LineSegment3f &lineInGadgetSpace, Gaff
 	const V2f rasterPosition = viewportGadget->gadgetToRasterSpace( lineInGadgetSpace.p1, this );
-	const V2f rasterPositionCenter = rasterPosition + V2f( 0.5f );
-	const LineSegment3f offsetLineInGadgetSpace = viewportGadget->rasterToGadgetSpace( rasterPositionCenter, this );
+	const LineSegment3f offsetLineInGadgetSpace = viewportGadget->rasterToGadgetSpace( rasterPosition, this );
```
Fixed in https://github.com/GafferHQ/gaffer/pull/5630/commits/2f459dd9df7a11d379922cd9b68d377a737097d9

ericmehl commented on src/GafferSceneUI/SceneGadget.cpp:
```diff
@@ -657,8 +657,8 @@ std::optional<V3f> SceneGadget::normalAt( const IECore::LineSegment3f &lineInGad
 	V2f centerRasterP = viewportGadget->gadgetToRasterSpace( lineInGadgetSpace.p0, this );
-	centerRasterP.x = round( centerRasterP.x );
-	centerRasterP.y = round( centerRasterP.y );
+	centerRasterP.x = (float)( (int)centerRasterP.x + 0.5 );
```
Fixed in https://github.com/GafferHQ/gaffer/pull/5630/commits/009cb96d41b768966e6631c319256efc6041aaa4

ericmehl commented:
I squashed the previous fixups and pushed three new ones to address the latest comments. I did a fresh round of testing and it still seems nice and accurate. Hopefully ready for the old final-final-for-real-this-time look!

johnhaddon reviewed (approved):
LGTM.

murraystevenson commented:
Thanks Eric! Merging.
